pythoncode/usbTrainer.py

```diff
@@ -92,486 +93,713 @@
-        Px      = P * factor
+        if self.UserAndBikeWeight < 70:
+            self.UserAndBikeWeight = 75 + 10
 
-        #-----------------------------------------------------------------------
-        # Translate power to resistance at the current speed
-        # Note that Power2Resistance() has its own limitations
-        #-----------------------------------------------------------------------
-        rtn     = Power2Resistance(Px, SpeedKmh, Cadence)
+        self.TargetPower = self.__PfietsNL()
 
         if debug.on(debug.Application):
-            logfile.Write ("%4.1f%%, P=%3.0fW(%3.0fW), R=%4.0f f=%4.2f" % \
-                     (TargetGrade, P,     Px,        rtn,  factor) )
-
-    rtn = ResistanceAtLowSpeed(SpeedKmh, rtn)
-
-    return rtn
+            logfile.Write ("%4.1f%%, P=%3.0fW, R=%4.0f" % \
+                (self.TargetGrade, self.TargetPower, self.TargetResistance) )
 
-def PfietsNL(TargetGrade, UserAndBikeWeight, SpeedKmh):
+    def __PfietsNL(self):
         #-----------------------------------------------------------------------
         # Thanks to https://www.fiets.nl/2016/05/02/de-natuurkunde-van-het-fietsen/
         # Required power = roll + air + slope + mechanical
         #-----------------------------------------------------------------------
         c     = 0.004                           # roll-resistance constant
-        m     = UserAndBikeWeight               # kg
+        m     = self.UserAndBikeWeight          # kg
         g     = 9.81                            # m/s2
-        v     = SpeedKmh / 3.6                  # m/s       km/hr * 1000 / 3600
+        v     = self.SpeedKmh / 3.6             # m/s       km/hr * 1000 / 3600
         Proll = c * m * g * v                   # Watt
 
         p     = 1.205                           # air-density
         cdA   = 0.3                             # resistance factor
         w     =  0                              # wind-speed
         Pair  = 0.5 * p * cdA * (v+w)*(v+w)* v  # Watt
 
-        i     = TargetGrade
+        i     = self.TargetGrade
         Pslope= i/100 * m * g * v               # Watt
 
         Pbike = 37
 
         return Proll + Pair + Pslope + Pbike
-
+        
+    #-------------------------------------------------------------------------------
+    # Convert WheelSpeed --> Speed in km/hr
+    # SpeedScale must be defined in sub-class
+    #-------------------------------------------------------------------------------
+    # WheelSpeed = 1 mm/sec <==> 1.000.000 / 3600 km/hr = 1/278 km/hr
+    # TotalReverse: Other implementations are using values between 1/277 and 1/360.
+    # A practical test shows that Cadence=92 gives 40km/hr at a given ratio
+    #		factor 289.75 gives a slightly higher speed
+    #		factor 301 would be good (8-11-2019)
+    #-------------------------------------------------------------------------------
+    def Wheel2Speed(self):

```

Comment:
I tend to dislike functions that have a side-effect (setting a member variable) where unnecessary. I saw where this function was called and immediately went "hold on... what's this doing?" Obviously I knew, but I would take WheelSpeed as an argument and return a value - and store to SpeedKmh at the calling site. It's a simple function, but it's very "coupled" the way it's written here. Splitting hairs - just an opinion.

Reply:
agree on the principle

But these functions are not generally callable but intend to calculate member-variables into member-variables and are a function to separate code into pieces.

It avoids passing lists of variables and returning a value to be put into a variable for the sole purpose of the principle.

Reply:
One could ask if it is even meaningful to have WheelSpeed as a member variable? The units we get in the response from the trainer are essentially of no use to the external interface of the class. And if the scale factor is a constant, then this function can become a class @staticmethod, not an instance method - preferring to avoid any reference to "self" if none is needed.

Reply:
We do not disagree on the basics, I think.

I read "self.function()" as function does an operation on the object self and hence modifying the attributes of self is not a side effect but the intended operation.

And re WheelSpeed; I have put the relevant attributes together in clsTacxTrainer where this one is arguable especially since not needed for iVortex. Having all relevant parameters together has it's advantage. I have decided to keep WheelSpeed to stop converting up-and-down but have WheelSpeed, SpeedKmh and VirtualSpeedKmh simultaneously available.


PS. VirtualSpeed must be based upon CurrentPower, not TargetPower and I will update that. The overall idea works, the formula's are not yet 100%.

cu!

Reply:
Such a simple function to debate. ;)

Re: VirtualSpeed, I think we are in slightly different directions with our respective ideas. I'm not quite understanding how speed can be worked out from grade and power though... are you assuming a steady state? I could be going much faster but putting in less power because I'm slowing down due to fatigue - I think you'd need to account for inertia in this case, too? Which implies some knowledge of current speed and therefore the calculation is calculating a projected speed? I'm getting too wrapped up.

My thoughts have been simply to have a deterministic (but potentially non-linear) relationship between wheel speed and (virtual) speed, and nothing more. I believe that concept alone solves all remaining issues around low-speed-high-load, running out of gears at high speed, accurate simulation, etc.

I'm not thinking too hard on it now - will let things settle and revisit some other time.

Reply:
We return Power, Cadence, Heartrate, Speed to CTP (Rouvy/Zwift) where they know the grade we are riding and they display a virtual speed that definitely differs from the speed we communicate. For sure Heartrate is not used, most likely Cadence not either. Cyclist weight will be used as well

So CTP calculate Speed from Power/Grade/Weight: speed = f(Power, Grade, Weight).

We have a formula to calculate Power from given Speed: P = Grade2Power(Grade, Weight, Speed) which can be inverted to f(Power, Grade, Weight).

Note that Grade2Power() delivers TargetPower
and f(Power, Grade, Weight) should use CurrentPower (which currently is not the case)

The VirtualSpeed implementation in FortiusANT is not complete yet, but an excellent idea from your side to implement and it was not too difficult either and done a little bit sloppy (see above) to get it out yesterday; but generally matching Rouvy.

Apart from the CurrentPower to be used, a suggestion for a better formula would be welcome if the FietsNL would not be good enough. I think you had a link, please resend. Does that contain the formula's?


My statement regarding WheelSpeed: think of a Tacx Neo: it does not have a Wheel and by consequence no WheelSpeed and still CTP's have the notion of Speed. So WheelSpeed is not an input.

Thanks for assistence and feedback!

Reply:
This is a topic full of discussion points. :)

1) I didn't suggest the inverse function from power+grade to speed.. that was you. It's definitely interesting to play with physics, but I'm not yet convinced of value.

2) If we set Zwift to 100% trainer difficulty so that it transmits correct grade, and we model required power precisely, accounting for ALL the parameters sent to us by Zwift over ANT+ (rolling resistance, drag coefficients, drafting factor, rider+bike weight, etc...), and if we have an accurate wheel2speed factor, and Fortius implements virtual flywheel in the same way as Zwift may or may not also model inertia, then there is no reason that FortiusANT speed should differ at all from Zwift. At all. It's physics. It's a model, but it's a well defined formula with known inputs. If the speeds differ, there will be a good (defineable, but perhaps hard to find) reason.

3) I don't believe Rouvy models inertia - grade changes translate to fairly sharp speed changes in-software. This is independent of whether FortiusANT models inertia or not - that's just down to whether you set VirtualFlywheel value or not.

4) A Tacx Neo may not have a wheel, but it has tooth counts, gear ratios and it has a rear-wheel-spindle speed. The only thing that is missing is a wheel diameter, then you can calculate a speed. Indeed you can simulate any wheel diameter. The Fortius has a wheel, and that wheel diameter is encoded into the Wheel2Speed() factor... change that factor, you can simulate the same gearing for a different wheel size. Wheel size is available on ANT+ where CTP transmits it... therefore Wheel2Speed() could well be parameterised by wheel size, with 700c x 25 and perhaps more specifically, the rolling circumference of the Blue TACX trainer tyre being the "sensible default". 


So... FietsNL Grade->Power()... the page I referenced with formula was https://www.gribble.org/cycling/power_v_speed.html. But they're standard formula, any number of sources are available. Worth noting that your fietsnl implementation makes a (reasonable) simplification to eliminate a few trigonometric functions but that's not going to make much difference - for purity, I'd put them back in myself. However, I think honoring the ANT+ track/wind coefficients is a more a priority... ie when you draft, you draft - same gear, same cadence, less power. ;) Aside - I also think that a pie chart or time-based stacked graph would be interesting... to show the relative (or absolute) contributions to required power of rolling resistance, wind resistance and grade/gravity.


Regarding VirtualSpeed... I need a different term for my idea. As I talked about wheel size and wheel2speed factor above, worth noting that conceptually, my "speed-to-virtual-speed" relationship could also be implemented as a non-constant-but-speed-dependent factor in Wheel2Speed()... whatever is most elegant is how it should be expressed, and I haven't yet come up with anything more elegant in concept than what is on my branch - the expression of the formula stinks I admit. But my point is that I don't need power to calculate my speed - I have a motor that tells me speed for that... all I need to do is to ensure that I can overcome low-speed-high-load... we do that by "allowing the wheel to spin faster for a given load" - that's done by giving the trainer a lower resistance value - that's done by disconnecting "actual wheel speed" from "simulated speed" through my deterministic-but-non-linear relationship. But it is a defined relationship - not an inverse power calculation - power only comes into the calculation of resistance to send to the unit, a function of target power and actual speed - simulated speed is simply and only a function of actual speed. So I don't know what to call it.

Interesting analogy perhaps - i-Vortex has a concept of VirtualSpeed... that if you send it a Power too high or too low for the current speed (?) it will change something and report a virtual speed instead. Do you understand it's behaviour in this respect?

This is all terribly important to me, it seems. I don't care what the physics model internal to some 3rd party model thinks, I can what FortiusANT thinks... given a Grade and all the other ANT+ parameters... ideally speed, power, intertia, etc... would be absolutely correct, 1:1, for my gearing and wheel size. FortiusANT has one limitation (high-load-low-speed, and I'm seeking to treat that specifically, all else remaining untouched, with the simplest conceptual addition possible. Right now I can't see past my speed curve relationship, but calculating virtual speed from power is not where I see it.

Reply:
![image](https://user-images.githubusercontent.com/15227658/81670663-dc69e780-943f-11ea-93ff-ce6b0f1cd962.png)
https://www.manualslib.com/manual/1157117/Tacx-I-Vortex.html?page=6#manual

Reply:
Thanks for actively thinking on this subject.

Immagine I ride on a flat surface and I produce 234Watt (just to name a figure). You would immagine that results in a certain speed. If I add power, speed will increase; my physical moving over the road and (let's forget about slippage) the rotation of the wheel as well.

It is impossible to put 350 Watts in my bike without going faster.
And it's impossible to keep the same speed reducing the power put in the bike.
** Assuming I do not touch my brake ** :-)


Being on a stationary trainer, I can apply 100Watts, 200Watts, 300Watts and never move; yes the wheel rotates but the force is put **into the brake**. (ERGmode).

In GradeMode I can ride uphill 10% 20kmh(wheelspeed) applying 100Watts; just reduce the resistance enough. This is physically impossible in the real world (I cannot reduce gravity).

In my opinion the question is:
How do Zwift/Rouvy calculate the speed displayed when we are riding?
And how do we equal that to display a fairly similar virtual speed on our screen?
The formula I now use is incorrect; I will study to improve.


"FortiusANT has one limitation (high-load-low-speed, and I'm seeking to treat that specifically" as you know I have been studying this since december and came to the PowerCurve solution which is too complex.
Discussing the subject a friend of mine came with the Virtual Gearbox idea and that works; it does exactly what we need.

Give me some time to get the formula's consistent. Ignore the displayed speed (which is not correct now) but do some tests in Grade mode and check what the Up/Down buttons with the Gearbox bring to you.

Reply:
Only if you maintain 234 Watts for a given amount of time will you result in attaining a certain theoretical speed. I think the your formula probably works "instantaneously" if you ignore inertia completely. But that wouldn't be a realistic simulation. So given power + given grade does not imply given speed. In the limit, yes, but it's an over-simplified model.

An excellent example of why this model is over-simplified is to disable the virtual flywheel with your new (correct) Grade2Power() approach, and then simulate going from a 0% steady state ride straight into a -5% downhill... the wheel speed almost immediately results in terminal velocity for that downhill. That's not a realistic simulation and is the reason that I wanted virtual flywheel value to be accurate a few weeks ago. You need to set the virtual flywheel value to achieve the gradual increase in speed that you would in real life. You could, alternatively set virtual flywheel to 0 and instead of letting the Fortius do it, model inertia in software yourself and gradually have the resistance requested of the trainer approach the theoretical value for a given weight+grade+speed, but this is more complex.

In terms of calculating in-game speeds, Zwift/Rouvy probably do just as you're trying to do... Zwift probably models inertia and so takes grade, reported power AND last known speed into account, I don't think Rouvy does model inertia, so is probably just a function of grade/power (and unrealistic for that reason). But I don't think FortiusANT needs to do this power->speed calculation at all. At least not unless implementing inertia in software for a trainer that doesn't have it, or perhaps trying to re-calculate a realistic equivalent real-world speed where a grade modifier may be in place for convenience or pleasure. Otherwise, you've "got the wheel already" - the calculation is not necessary.

I believe that if you are trying to ride an unrealistic grade/speed, then in the purest sense, I don't think you should be able to - because it's not a realistic simulation - or to put it another way, it should be possible to not be able to - you should be able to mash the pedals at massive Watts trying and failing up the steepest ascent, despite Fortius low-speed-high-load limitation, because it would be a realistic simulation to attempt and fail. There are two exceptions to this belief in total realism... one is for convenience/pleasure, and one is a workaround for Fortius limitation.

First exception is a grade modifier... pretend that 20% is only 10% and you'll get up it, and this is probably where your inverse calculation (plus inertia) would come in to determine what "real-world" is again - you don't want your ride to become un-rideable - you need a way to see past realism and just get on and ride. That's Zwift's "trainer difficulty setting".

The other exception is along the same lines of "let it be ridden, despite it not being realistic" and is simply a consequence of my virtual speed relationship function and that is that you happen to gain unrealistic new low gears to enable you to crawl up the slope at 2kph - which is where an indication of "you've exceed your real-world bike gear-inches" would be useful - to know that I am no longer in touch with reality. No longer needing a grade modifier to climb unrealistic slopes, you now have unrealistic gears. (Aside: the other "benefit" of Zwift's grade modifier is that you require fewer gear changes on average, which pleases some. Not me.)

Unless implementing a grade modifier (which I'd argue is not necessary for any practical purpose), you don't need to use power and physics to calculate what the real-world simulated speed would be - the trainer's already simulating this for you. Ideally, it would be 1:1, but it can't be for steep grades, so I make the relationship non-linear and there's still no power calculating required - just a speed conversion function to and fro from real to simulated. I don't think it could be implemented in any simpler terms. Only with a grade modifier, if you wish to report an adjusted simulated speed that corresponds to what it would be for the pre-modified grade then it gets trickier, or you ignore speed.

I will have a ride shortly, but I'm still firmly on my side of the fence. :)

Reply:
Before I ride, I'm starting to understand your powercurve factor.

From Grade and Speed you work out Power required. Then you multiply that Power by some PowerFactor (changeable using Up/Down buttons), and then convert that Factor-modified-Power to Resistance.

What is the PowerFactor achieving? It's artificially increasing or decreasing the required Power that has been calculated for the current Speed at the current Grade. Meaning that if you want to maintain the same Cadence and not change RealGear, you have to increase/decrease Power. So while the ActualSpeed remains the same, the Power applied changes and therefore the VirtualSpeed changes. This is therefore a VirtualGear. At low speeds and high loads you can therefore change into a lower virtual gear which will drop the required Power, drop the Resistance value and result in some lower VirtualSpeed. And you need to get back from the CurrentPower to the VirtualSpeed?

Is this the way you look at it?

Reply:
OK - I've ridden. First lesson was which way round the head unit goes on the handlebars - text has rubbed off. Up was down.

I think I get the powercurvefactor. Simply put, it reduces resistance in order to make it feel less/more difficult - as if you were riding a smaller or bigger gear, at the expense that wheel speed is no longer realistic to the simulation, and in a not-easy-to-reverse-the-calculation way. Though 1.1^N as powercurve factor seems rather arbitary - does the tooth count actually mean something or is it just a reference point? I am running on Raspberry PI so I can't get the GUI and I don't see the gear icon.

The virtual speed calculation is obviously doing something good, but there are issues, like it sticking at 49 km/h until a grade change or something like that, or I can be screaming down hill and stop pedalling and my virtual speed drops pretty quickly due to drop in power (no inertia!). But I get it.

Still, I think it's not the most elegant way in which to achieve the goal.

TargetPower is a function of Speed and Grade.
- Your approach modifies the output, TargetPower
- A grade modifier (eg Zwift's trainer difficulty) modifies the Grade input
- my approach modifies the Speed input

They're all aiming at the same thing, but in different ways.

With Grade or TargetPower, once modified, wheel speed is somewhat disconnected with the simulation and it gets complicated to work out a VirtualSpeed and you need inverse physics. The thing with Speed is it's two way - you can easily convert back with a very simple conversion. Whether conceptually it's a Speed-to-VirtualSpeed function like I started out with, or simulating different wheel diameters, it still feels special. Unless I'm really missing something, and I might be, which would be embarrassing given how much thought I've put to the idea, and how convinced I appear to be. :/

Consider up/down buttons changing the virtual wheel diameter. Conceptually it's slightly odd, but I think it's a valid approach to solving the problem, and cleaner than inverse physics. Worth working it through as a thought exercise.

If I'm going up an unachievable slope and drop down to 20" wheels, then I travel less distance for a full wheel revolution than I would on 700c wheels. The bike+trainer obviously still has a 700c wheel, but by using a smaller virtual wheel, we gain more low end gears by simulating a slower virtual speed using a higher actual speed, and can easily convert between the two. Or by increasing wheel size we can gain top end gears. The buttons do essentially the same thing in that they make it easier/harder to pedal, but to report VirtualSpeed, no inverse physics is needed.

I'm talking in terms of wheel size only because it's a convenient concept - it's effectively the same thing as a speed-relationship. And happy to ignore non-linearity for now and focus on the idea of button pushes being what alters the feel. Changing a wheel size basically changes the gradient of the speed relationship, but it remains linear. If wheel size were speed-dependent, then it would be non-linear, and starting to get interesting.

I wonder if I'm making sense.

Reply:
> Though 1.1^N as powercurve factor seems rather arbitary
Not really, a cassette 10-11-12-14...36-40 has a 10% step, limitted by the fact that you have to work with whole numbers of teeth. Inspired by that.

First I displayed the factor, which is quite meaningless. Then I display the "cassette" and yes - it's a reference. Bigger number is lighter, lower number is heavier. No rider on his bike is calculating when shifting gears; actually you do the same and feel the effect.

Shimano and friends are trying to make life easier with 8-9...13 gears. I think FortiusANT is the first bike with ... I do not even know how many steps.

> The virtual speed calculation
Sorry, the VirtualSpeed calc is crap. Working on a better algorithm. 
Note that P = v**3 + v**2 + v + c is not easily inversable to v = f(P) especially because mathematically there are 3 answers.


I think you suggestion for a Virtual Speed was great and implicitly asked by others. For now I continue the approach to get the implementation working; different implementations can be done always lateron.

> Consider up/down buttons changing the virtual wheel diameter
I understand what you mean but in the world where 10% of the users do not even know how to make a commandline, it is overasking the community in understanding what it is.


The concept now is:
- CTP gives grade
- FortiusANT calculates required Power -> Resistance
- Tacx returns currentResistance -> currentPower
- FortiusANT returns currentPower to CTP

When PowercurveFactor = 1, changing gears on the physical bike is the way to adjust to demand. If Grade2Power() matches how CTP looks at it; Zwift/Rouvy will display the same speed value as FortiusANT (the speed returned by the trainer).

There are three ways to decrease the power to be applied:
- Reduce cadence: same resistance, less rpm = less power
- Gears on the bike: reduce resistance, same rpm = less power
- Down button from FortiusANT: reduce resistance, same rpm = less power

The difference between 2 and 3 is that the wheelspeed remains higher in 3 which makes life easier for the Fortius.
Of course all combinations between 1/2/3 are possible.

Since it's in development, I finish first. Then test. And get feedback on the implementation.

Reply:
All good.

You might save yourself effort to scrap VfietsNL now and look at simply using your powercurvefactor against self.SpeedScale, leaving the buttons doing what they're doing.

Calculations are:
- TargetPower as a function of TargetGrade and **VirtualSpeed**
- TargetResistance as a function of TargetPower and **ActualSpeed**
- CurrentPower as a function of CurrentResistance and **ActualSpeed**
- **VirtualSpeed** as a function of **ActualSpeed** and your factor.

Ironic how this discussion is on Wheel2Speed().

Reply:
I must say, you are fast.

Spot on, I like this solution very much. Thank you so much. I could just discard some hours of working in another direction, trying to find the Zwift/Rouvy formula, which I do not need since I know the virtual gear I'm in.

Reply:
@WouterJD I am happy you like the solution.

It was difficult to express in words, or find a way to describe it, or a real-word concept to relate it to, but the key is that for the reverse calculation, you are right - you already know the "virtual gear". I stumbled upon the idea and then the effort was in evaluating it. Sorry for writing so many words but I think it's a really great point to have reached in the software!

Feedback on it...

- 0.3 is not sufficient for low-gearing... on a 20% at 90kg+ I find myself needing more. Suggest 0.1 (for 10x) or 0.2 (for 5x) as lower limit? Upper limit as 3 is probably ok, haven't really tried.
- Personally, I don't want to touch the buttons much -  I may encode a non-linear relationship into the speedscale factor so I get, hands-off, extra low, extra high, and linear middle as I had played with previously, but that said, not wanting to touch buttons _too_ much but accepting it is the interface, I imagine the use of it will be in slight panic in-game and in response to an "oops... need more gear here quick..." then increasing the 1.1 to 1.5 might be better... meaning only 3-5 presses for hard uphills, 2-3 presses for screaming downs. 
- since I am running without GUI - currently, during development I guess, it would be useful to see the factor, or the equivalent wheel diameter, or whatever metric it makes sense to display. If the factor, then 2 decimal places perhaps, though if you decide on a 1.5 step value instead of 1.1, then perhaps just showing discrete "levels"... eg -1, -2, -3, +1, +2, +3 might make more sense. I don't know.
- Basically - GREAT! Maths is simpler, true/known relationship to wheel speed, no guesswork, low-end-high-load is worked around, top end has more gears when needed, and tuning/reporting could do with a tweak

Personally I'm thinking of trying 1.5 multiplier with a 5 step limit down (factor 0.132)  and a 4 step limit up (factor 5.06).

Then, given we _know_ that the low end suffers, there might be no harm in choosing a non-linear relationship in some upcoming release... I don't know. Need some kms/% under our belts and some feedback from others.

I'm excited. Well done, Thanks, Etc.

Reply:
Feedback on my ride last night...

- having not-too-many-but-reasonably-large jumps for "button down" makes sense... eg 1, 0.7, 0.45, 0.3, 0.2 or similar.
- having fewer (smaller) jumps for "button up" makes sense - 1.5x per-button-push is too large, presumably because air resistance has such a MASSIVE impact as you increase speed... so 1, 1.25, 1.5, 1.75 probably enough.


However... even a modest grade on Zwift (with trainer difficulty 100%), riding Yorkshire with the odd 6% or 7% grade climb at 95-kg or so, I was struggling at low-end.

This definitely suggests to me that the default 1:1 power factor at low speed is wrong... we should probably default in our implementation to automatically become non-linear at lower speeds without any user-interaction... ie as you slow down, the factor drops <1 without user having to push any buttons... we shouldn't HAVE to push buttons for modest grades, we should probably at least try to mask the limitation out of the box, if not all the way down to 20%, then certainly for 7%. I don't know if there's a "VirtualSpeed" flag we can set in the FE broadcast to indicate that we're doing tricksy things with the power/speed... I guess that's how i-Vortex uses its VirtualSpeed flag.

Reply:
And regarding the buttondown factor; a shimano cassette has a 10% step (10-11-12....36-40) doing big steps will trigger the inverse critique.

I get Grade, calculate power from it.
If you are on a slope and you wanna ride away, start in a low gear.

What does "I was struggling at low-end" mean?
What required power was displayed when doing 7%